cpp/src/io/parquet/reader_impl.cpp

```diff
@@ -122,7 +156,13 @@ void reader::impl::decode_page_data(size_t skip_rows, size_t num_rows)
   chunk_nested_valids.host_to_device_async(_stream);
   chunk_nested_data.host_to_device_async(_stream);
 
+  // TODO: explore launching these concurrently with a stream pool
   gpu::DecodePageData(pages, chunks, num_rows, skip_rows, _file_itm_data.level_type_size, _stream);
+  if (has_strings) {
+    chunk_nested_str_data.host_to_device_async(_stream);
```

Comment:
So I have one last request.  The lingering concern here is running these separate kernels on the same stream.  In situations where we have small-ish files with a mixed number of types - strings and ints, running these kernels on the same stream potentially impacts performance seriously.  

If we have a GPU with 100 SMs, and we have 50 string pages and 50 int pages, normally they would all run at once.  But with this new arrangement, 50 would run, then another 50, halving gpu utilization.

The existing cudf benchmarks are not going to expose this because they favor generating inputs with huge numbers of pages, so the GPU will always be saturated.  So what I'd like to see is this:

- Add a benchmark that creates a reasonably large file with a similar to above profile.  Strings, ints.  40 pages of each (a T4 has 80 SMs so this would be a nice universal test).  You should be able to force this arrangement with the writer via `.max_page_size_bytes` and `.max_page_size+rows`

- Do a before/after comparison with your stuff.  There should be a pretty clear difference.

- As a step 2, at the top of decode_page_data, add this little static initializer for a stream pool, and then pull a stream out of the pool to invoke the string kernel via `.get_stream` (make sure to synchronize on this stream at a rational point alongside the core stream.  
```
auto& get_stream_pool(){
   static auto pool = rmm::cuda_stream_pool(10);
    return pool;
}
```

There should be a clear performance win (including some clear differences in an nsys trace).  Based on some results I've had with recent optimization attempts, running multiple, type-specific kernels is something we might be interested in pursuing further.  This string stuff might be a nice guinea pig.

Reply:
Pushed some changes...still working on some.  Results so far:
```
# parquet_read_io_small_mixed without stream pool

|    io    |  cardinality  |  run_length  |  num_string_cols  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |       Diff |   %Diff |
|----------|---------------|--------------|-------------------|------------|-------------|------------|-------------|------------|---------|
| FILEPATH |       0       |      1       |         1         |   4.925 ms |       5.29% |   5.005 ms |       1.74% |  79.316 us |   1.61% |
| FILEPATH |     1000      |      1       |         1         |   2.307 ms |       2.72% |   3.356 ms |       2.53% |   1.049 ms |  45.45% |
| FILEPATH |       0       |      32      |         1         |   4.749 ms |       4.27% |   4.794 ms |       1.49% |  45.341 us |   0.95% |
| FILEPATH |     1000      |      32      |         1         |   1.357 ms |       2.93% |   1.942 ms |       6.34% | 585.435 us |  43.15% |
| FILEPATH |       0       |      1       |         2         |   5.483 ms |       1.72% |   5.516 ms |       2.80% |  32.711 us |   0.60% |
| FILEPATH |     1000      |      1       |         2         |   2.308 ms |       2.81% |   3.356 ms |       2.33% |   1.049 ms |  45.44% |
| FILEPATH |       0       |      32      |         2         |   5.267 ms |       1.82% |   5.304 ms |       1.44% |  37.500 us |   0.71% |
| FILEPATH |     1000      |      32      |         2         |   1.469 ms |       2.24% |   1.937 ms |       1.46% | 468.230 us |  31.88% |
| FILEPATH |       0       |      1       |         3         |   5.948 ms |       1.39% |   5.983 ms |       1.22% |  35.133 us |   0.59% |
| FILEPATH |     1000      |      1       |         3         |   2.338 ms |       7.68% |   3.362 ms |       2.08% |   1.023 ms |  43.78% |
| FILEPATH |       0       |      32      |         3         |   5.797 ms |       2.04% |   5.809 ms |       1.24% |  11.800 us |   0.20% |
| FILEPATH |     1000      |      32      |         3         |   1.583 ms |       2.43% |   1.934 ms |       1.53% | 351.543 us |  22.21% |

# parquet_read_io_small_mixed with stream pool

|    io    |  cardinality  |  run_length  |  num_string_cols  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |        Diff |   %Diff |
|----------|---------------|--------------|-------------------|------------|-------------|------------|-------------|-------------|---------|
| FILEPATH |       0       |      1       |         1         |   4.925 ms |       5.29% |   4.935 ms |       4.67% |    9.449 us |   0.19% |
| FILEPATH |     1000      |      1       |         1         |   2.307 ms |       2.72% |   2.267 ms |       3.16% |  -39.907 us |  -1.73% |
| FILEPATH |       0       |      32      |         1         |   4.749 ms |       4.27% |   4.757 ms |       3.02% |    8.061 us |   0.17% |
| FILEPATH |     1000      |      32      |         1         |   1.357 ms |       2.93% |   1.532 ms |       2.18% |  175.445 us |  12.93% |
| FILEPATH |       0       |      1       |         2         |   5.483 ms |       1.72% |   5.204 ms |       8.58% | -279.011 us |  -5.09% |
| FILEPATH |     1000      |      1       |         2         |   2.308 ms |       2.81% |   2.263 ms |       3.27% |  -44.884 us |  -1.95% |
| FILEPATH |       0       |      32      |         2         |   5.267 ms |       1.82% |   5.217 ms |       1.94% |  -49.412 us |  -0.94% |
| FILEPATH |     1000      |      32      |         2         |   1.469 ms |       2.24% |   1.612 ms |       2.43% |  143.197 us |   9.75% |
| FILEPATH |       0       |      1       |         3         |   5.948 ms |       1.39% |   5.910 ms |       3.03% |  -37.821 us |  -0.64% |
| FILEPATH |     1000      |      1       |         3         |   2.338 ms |       7.68% |   2.268 ms |       3.91% |  -69.983 us |  -2.99% |
| FILEPATH |       0       |      32      |         3         |   5.797 ms |       2.04% |   5.862 ms |       1.54% |   65.267 us |   1.13% |
| FILEPATH |     1000      |      32      |         3         |   1.583 ms |       2.43% |   1.625 ms |       1.60% |   42.474 us |   2.68% |
```
As expected times are bad for the cases where the GPU is left half idle because of the serial nature of decoding the pages.  Using a stream pool to overlap the decoding kernels does much better (but still a little worse than the old code).

Reply:
![Screenshot from 2023-06-21 11-40-00](https://github.com/rapidsai/cudf/assets/25541553/702590e5-a86f-43e5-8a28-80958ff86753)

Here's the profile from the 1000/32/2 benchmark.

Reply:
I tried to streamline `setupLocalPageInfo` as @nvdbaranec suggested offline, but there's some wrinkle I'm missing. One of the unit tests fails with null counts larger than expected. I'll give this another go.

Reply:
@vuule can you look at how I did the stream pool?  I wound up adding it to the impl object rather than using a static one.

Reply:
latest benchmark with the stream pool in place:
```
# parquet_read_decode

|  data_type  |      io       |  cardinality  |  run_length  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |         Diff |   %Diff |
|-------------|---------------|---------------|--------------|------------|-------------|------------|-------------|--------------|---------|
|  INTEGRAL   | DEVICE_BUFFER |       0       |      1       |  14.446 ms |       0.75% |  14.836 ms |       0.99% |   390.704 us |   2.70% |
|  INTEGRAL   | DEVICE_BUFFER |     1000      |      1       |  13.672 ms |       0.44% |  14.000 ms |       0.50% |   327.931 us |   2.40% |
|  INTEGRAL   | DEVICE_BUFFER |       0       |      32      |  12.078 ms |       0.50% |  12.566 ms |       0.48% |   488.227 us |   4.04% |
|  INTEGRAL   | DEVICE_BUFFER |     1000      |      32      |  11.556 ms |       0.45% |  12.051 ms |       0.50% |   494.633 us |   4.28% |
|    FLOAT    | DEVICE_BUFFER |       0       |      1       |   8.755 ms |       0.38% |   8.772 ms |       0.36% |    16.405 us |   0.19% |
|    FLOAT    | DEVICE_BUFFER |     1000      |      1       |   8.589 ms |       0.50% |   8.545 ms |       0.53% |   -44.250 us |  -0.52% |
|    FLOAT    | DEVICE_BUFFER |       0       |      32      |   7.287 ms |       0.55% |   7.290 ms |       0.70% |     2.525 us |   0.03% |
|    FLOAT    | DEVICE_BUFFER |     1000      |      32      |   7.549 ms |       1.39% |   7.544 ms |       2.35% |    -4.273 us |  -0.06% |
|   DECIMAL   | DEVICE_BUFFER |       0       |      1       |  19.095 ms |       0.50% |  18.840 ms |       0.75% |  -254.786 us |  -1.33% |
|   DECIMAL   | DEVICE_BUFFER |     1000      |      1       |   7.569 ms |       0.65% |   7.501 ms |       0.58% |   -67.905 us |  -0.90% |
|   DECIMAL   | DEVICE_BUFFER |       0       |      32      |   9.118 ms |       0.77% |   9.116 ms |       0.50% |    -1.569 us |  -0.02% |
|   DECIMAL   | DEVICE_BUFFER |     1000      |      32      |   6.291 ms |       0.71% |   6.230 ms |       0.67% |   -61.431 us |  -0.98% |
|  TIMESTAMP  | DEVICE_BUFFER |       0       |      1       |  25.775 ms |       0.57% |  25.213 ms |       0.50% |  -561.543 us |  -2.18% |
|  TIMESTAMP  | DEVICE_BUFFER |     1000      |      1       |   8.146 ms |       0.57% |   8.046 ms |       0.49% |   -99.693 us |  -1.22% |
|  TIMESTAMP  | DEVICE_BUFFER |       0       |      32      |  10.686 ms |       1.56% |  10.580 ms |       0.95% |  -106.155 us |  -0.99% |
|  TIMESTAMP  | DEVICE_BUFFER |     1000      |      32      |   7.087 ms |       0.57% |   7.018 ms |       0.71% |   -68.621 us |  -0.97% |
|  DURATION   | DEVICE_BUFFER |       0       |      1       |  10.709 ms |       0.50% |  10.771 ms |       0.50% |    62.400 us |   0.58% |
|  DURATION   | DEVICE_BUFFER |     1000      |      1       |   7.828 ms |       0.42% |   7.889 ms |       0.65% |    60.604 us |   0.77% |
|  DURATION   | DEVICE_BUFFER |       0       |      32      |   7.860 ms |       0.41% |   7.902 ms |       0.50% |    41.143 us |   0.52% |
|  DURATION   | DEVICE_BUFFER |     1000      |      32      |   6.859 ms |       1.31% |   6.897 ms |       0.54% |    37.835 us |   0.55% |
|   STRING    | DEVICE_BUFFER |       0       |      1       |  50.841 ms |       0.49% |  41.080 ms |       0.50% | -9760.994 us | -19.20% |
|   STRING    | DEVICE_BUFFER |     1000      |      1       |  17.094 ms |       0.59% |  13.336 ms |       0.63% | -3757.784 us | -21.98% |
|   STRING    | DEVICE_BUFFER |       0       |      32      |  50.890 ms |       0.68% |  41.009 ms |       0.50% | -9880.564 us | -19.42% |
|   STRING    | DEVICE_BUFFER |     1000      |      32      |  15.565 ms |       0.24% |  12.407 ms |       0.71% | -3157.489 us | -20.29% |
|    LIST     | DEVICE_BUFFER |       0       |      1       |  80.768 ms |       0.40% |  80.828 ms |       0.20% |    60.150 us |   0.07% |
|    LIST     | DEVICE_BUFFER |     1000      |      1       |  72.315 ms |       0.17% |  72.743 ms |       0.28% |   428.268 us |   0.59% |
|    LIST     | DEVICE_BUFFER |       0       |      32      |  62.031 ms |       0.29% |  62.582 ms |       0.36% |   551.235 us |   0.89% |
|    LIST     | DEVICE_BUFFER |     1000      |      32      |  62.994 ms |       0.20% |  63.540 ms |       0.23% |   546.548 us |   0.87% |
|   STRUCT    | DEVICE_BUFFER |       0       |      1       |  63.745 ms |       0.26% |  56.472 ms |       0.22% | -7273.554 us | -11.41% |
|   STRUCT    | DEVICE_BUFFER |     1000      |      1       |  37.924 ms |       0.37% |  34.097 ms |       0.32% | -3827.127 us | -10.09% |
|   STRUCT    | DEVICE_BUFFER |       0       |      32      |  61.452 ms |       0.61% |  54.311 ms |       0.68% | -7140.755 us | -11.62% |
|   STRUCT    | DEVICE_BUFFER |     1000      |      32      |  35.541 ms |       0.36% |  31.730 ms |       0.14% | -3810.204 us | -10.72% |
```